Add @byrow attempt 2 (#250)
JuliaData/DataFramesMeta.jl: docs/src/index.md

  convenient syntax
* `@eachrow` and `@eachrow!` for looping through rows in data frame, again with high performance and 
  convenient syntax. 

> * `@byrow` for applying functions by-row of a data frame.

Reviewer: ```suggestion
* `@byrow` for applying functions to each row of a data frame (only supported inside other macros).
```

end
```

## Row-wise transformations with `@byrow`

> DataFrames provides the function-wrapper `ByRow`. `ByRow(f)(x, y)`

Reviewer 2: I am not sure what style you follow in DataFramesMeta.jl, but in DataFrames.jl we always add .jl to package names

Author: added `.jl` but will make a point to do this before 1.0.

is roughly equivalent to `f.(x, y)`, with a few exceptions discussed below. 
DataFramesMeta allows for users to construct expressions using `ByRow` 

> function wrapper with the syntax `@byrow`. 

Reviewer 2: maybe highlight that it is not a macro?

Reviewer 2: ah - you did below :)

```julia
@transform(df, y = @byrow :x == 1 ? "true" : "false)
```

becomes

```
transform(df, :x => ByRow(x -> x == 1 ? "true", "false") => :y)
```

!!! note 

>     Unlike `@.`, `@byrow` is not a "real" macro and cannot be used outside of 

Reviewer 2: the question is what are the parsing rules for `@byrow`? does it take exactly one expression that follows it?

Author: I included a more detailed discussion in updated docs. Hopefully it's clear.

    DataFramesMeta macros. However it's behavior within DataFramesMeta 
    macros should be indistinguishable from externally defined macros. 

### Comparison with `@eachrow`

In previous versions of DataFramesMeta, `@eachrow` was named `@byrow`. 

> This version of `@byrow` is deprecated, but the syntax can be used

Reviewer 2: "the syntax" - which syntax?

Author: clarified.

> to for similar, but not identical, behavior.

Reviewer 2: ```suggestion
too for similar, but not identical, behavior.
```

The syntax 

```julia
@eachrow df begin 
    :a * :b
end 
```

is similar to 

```julia
begin
    function tempfun(a, b)
        for i in eachindex(a)
            a[i] * b[i]
        end
    end
    tempfun(df.a, df.b)
    df
end
```

The function `*` is applied by-row. But the result of those operations

> is not stored in a new vector. Additionally, `@eachrow` and `@eachrow!`

Reviewer 2: where is it stored then? (or not stored unless stored explicitly?)

Author: it's not stored. It's literally just a `for`-loop. Hopefully the re-write is clear.

return data frames. 

By contrast,

```julia

> @with df @byrow begin 

Reviewer 2: this is confusing. I understand you refer to the deprecated `@byrow`? If this is the case then I would skip it - no need to document (or even mention deprecated functionality except for writing that it is deprecated).

If you refer to new `@byrow` then it is strange, as earlier you have written that `@byrow` is restricted to `@transform`.

Reviewer 2: Looking at your examples below it seems `@byrow` is accepted not only in `@transform` - I would suggest explicitly listing places where it is allowed.

Author: Added.

   2 │     2      4
```

### Comparison with `@.` and Base broadcasting

Base Julia provides the broadasting macro `@.` and in many cases `@.` 
and `@byrow` will give equivalent results. But there are important 
deviations in behavior. Consider the setup 

```julia
df = DataFrame(a = [1, 2], b = [3, 4])
```

* Control flow. In all versions of Julia, expressions of the form 
`if...else`, `a ? b : c` cannot be broadcasted. In versions below
1.7-dev, expressions of the form `a && b` and `a || b` cannot be 
broadcasted. Consequently, the `@.` macro will fail when encountering such
control flow while `@byrow` will not. 
```
julia> @with df @byrow begin 
           if :a == 1
               5
           else 
               10
           end
       end
2-element Vector{Int64}:
  5
 10

julia> @with df @. begin 
           if :a == 1
               5
           else 
               10
           end
       end # will error
```

* Broadcasting objects that are not columns. `@byrow` constructs an 
anonymous function *which accepts only the columns of the dataframe*
and broadcasts that function. Consequently, it does not broadcast
objects that are referenced which are not columns. 
```julia

> @with df @byrow :x + [5, 6]

Reviewer 2: will this always error? What if column `:x` would be a vector of two element vectors?

Author: Yes, I have clarified in the example.

```
will error. On the other hand
```julia

> @with df @. :x + [5, 6]

Reviewer 2: It will only work if `:x` has 2 elements - right?

Author: yes. I have added more detail.

```
will not. 

* Broadcasting expensive calls. In Base Julia, broadcastsing 
evaluates calls first and then broadcasts the result. Because
`@byrow` constructs an anonymous function and evaluates

> that function for every row in the DataFrame, expensive functions

Reviewer 2: ```suggestion
that function for every row in the data frame, expensive functions
```

will be evaluated many times. 
```julia
julia> function expensive()
           sleep(.5)
           return 1
       end;

julia> @time @with df @byrow :a + expensive();
  1.037073 seconds (51.67 k allocations: 3.035 MiB, 3.19% compilation time)

julia> @time @with df :a .+ expensive();
  0.539900 seconds (110.67 k allocations: 6.525 MiB, 7.05% compilation time)

```
This problem comes up when using the `@.` macro as well, but can easily be fixed with `$`.
```julia
julia> @time @with df @. :a + expensive();
  1.036888 seconds (97.55 k allocations: 5.617 MiB, 3.20% compilation time)

julia> @time @with df @. :a + $expensive();
  0.537961 seconds (110.68 k allocations: 6.525 MiB, 6.73% compilation time)
```
No such solution currently exists with `@byrow`. 

> 

Reviewer 2: as commented earlier - from the description it is not clear how `GroupedDataFrame` is handled with `@byrow` and I did not see any tests of `GroupedDataFrame` below.

Author: Added in the re-write. In general it's a bit pointless to use `@byrow` with a GroupedDataFrame as with `ByRow`. Though maybe if one wants to combine operations which take advantage of the grouping and operations which do not it could be useful.

Reviewer 2: Agreed - but the documentation should precisely state "how things work" (this is a definite reference for a contract of the functionalities you provide), and then also recommend "what makes sense" if needed I think.

end
```

## Row-wise transformations with `@byrow`

DataFrames.jl provides the function-wrapper `ByRow`. `ByRow(f)(x, y)`

> is roughly equivalent to `f.(x, y)`, with a few exceptions discussed below. 

Reviewer 2: do you discuss these exceptions indeed? The exceptions are:
* `AsTable` case - which I think we do not cover yet
* potentially another type of container can be allocated as we use `map` which is not guaranteed to be 100% consistent with broadcasting in corner cases

(I am not sure it is worth to go into such details in the documentation but I felt it is good to leave this comment)

Author: I added a short note about the broadcasting issue. 

I will avoid discussing `AsTable` until we implement that.

DataFramesMeta.jl allows users to construct expressions using `ByRow` 
function wrapper with the syntax `@byrow`. 

`@byrow` is not a "real" macro and cannot be used outside of 

> DataFramesMeta macros. However it's behavior within DataFramesMeta.jl

Reviewer 2: ```suggestion
DataFramesMeta.jl macros. However, it's behavior within DataFramesMeta.jl
```

macros should be indistinguishable from externally defined macros. 
Thought of as a macro `@byrow` accepts a single argument and 
creates an anonymous function wrapped in `ByRow`.  For example,

```julia
@transform(df, y = @byrow :x == 1 ? "true" : "false)
```

becomes

```julia
transform(df, :x => ByRow(x -> x == 1 ? "true", "false") => :y)
```

Macros that accept `@byrow`:

> * `@transform` and `@transform!`, `@select`, `@select!`, and `@combine`. `
>   @byrow` appears in the right hand side of expressions, of the form `@select(
>   df, z = @byrow :x * :y)`. 

Reviewer 2: ```suggestion
* `@transform` and `@transform!`, `@select`, `@select!`, and `@combine`.
  `@byrow` appears in the right hand side of expressions, of the form
  `@select(df, z = @byrow :x * :y)`. 
```

* `@where` and `@orderby`, with syntax of the form `@where(df, @byrow :x > :y)`
* `@with`, where the anonymous function created by `@with` is wrapped in

>   `ByRow`. `@with df @byrow :x * :y`, which is conceptually similar to 

Reviewer 2: for consistency with other examples:
```suggestion
  `ByRow`. `@with(df, @byrow :x * :y)`, which is conceptually similar to 
```

  `map((x, y) -> x * y, df.x, df.y)`.

To avoid writing `@byrow` multiple times when performing multiple
operations, it is allowed `@byrow` at the beginning of a block of 

> operations. All transformations in the block will operate by row.

Reviewer 2: one transformation is one expression?

Author: Yes, that's the new rule after #245. Hopefully it's not too confusing to people.

```julia
julia> @where df @byrow begin 
           :a > 1
           :b < 5
       end
1×2 DataFrame
 Row │ a      b     
     │ Int64  Int64 
─────┼──────────────
   1 │     2      4
```

`@byrow` can be used on macros which accept `GroupedDataFrame`s,
however, like with `ByRow` in DataFrames.jl, when `@byrow` is

> used, functions do not take advantage of the grouping, so the 
> behavior of `@transform(df, y = @byrow f(:x))` and 

Reviewer 2: I add this comment as with `@combine` the result would be different:
```suggestion
used, functions do not take advantage of the grouping, so 
for example the result of `@transform(df, y = @byrow f(:x))` and 
```

`@transform(groupby(df, :g), y = @byrow f(:x))` is the same.

### Comparison with `@eachrow`

> In previous versions of DataFramesMeta, `@eachrow` was named `@byrow`. 

Reviewer 2: ```suggestion
In previous versions of DataFramesMeta.jl, `@eachrow` was named `@byrow`. 
```

> The old macro `@byrow` is deprecated, but the `@byrow` syntac can 

Reviewer 2: ```suggestion
The old macro `@byrow` is deprecated, but the `@byrow` syntax can 
```

Author: No. By "`@byrow` syntax" I mean `@with df @byrow ...` is very similar to `@eachrow df ...`. In that it's looking through rows.

end
```

> ## Row-wise transformations with `@byrow`

Reviewer: I think the second paragraph still applies: it would be nice to start with a sentence or two saying that `@byrow` allows writing code that is applied to each row instead of having to vectorize it.

`@transform(groupby(df, :g), y = @byrow f(:x))` is the same.

### Comparison with `@eachrow`

In previous versions of DataFramesMeta.jl, `@eachrow` was named `@byrow`. 
The old macro `@byrow` is deprecated, but the `@byrow` syntax can 
now be used for similar, but not identical, behavior.

To re-cap, the `@eachrow` rougly transforms

```julia
@eachrow df begin 
    :a * :b
end 
```

to 

```julia
begin
    function tempfun(a, b)
        for i in eachindex(a)
            a[i] * b[i]
        end
    end
    tempfun(df.a, df.b)
    df
end
```

The function `*` is applied by-row. But the result of those operations
is not stored anywhere, as with `for`-loops in Base Julia. 
Rather, `@eachrow` and `@eachrow!` return data frames. 

Now consider `@byrow`. `@byrow` transforms

```julia
@with df @byrow begin 
    :a * :b
end
```

to 

```julia
tempfun(a, b) = a * b
tempfun.(df.a, df.b)
```

In contrast to `@eachroe`, `@with` combined with `@byrow` will return a vector of the broadcasted multiplication and not a data frame.

Additionally, `@eachrow` and `@eachrow!` allow modifying a data
data frame. Just as with Base Julia broadcasting, `@byrow` will

> not update columns. 

Author: This seems right.

> ```julia

Reviewer: How about making this a doctest?

Author: I will procrastinate this for a "jldoctest" PR.

julia> df = DataFrame(a = [1, 2], b = [3, 4]);

julia> @with df @byrow begin 
           :a = 500
       end
2-element Vector{Int64}:
 500
 500

julia> df
2×2 DataFrame
 Row │ a      b     
     │ Int64  Int64 
─────┼──────────────
   1 │     1      3
   2 │     2      4
```

### Comparison with `@.` and Base broadcasting

Base Julia provides the broadasting macro `@.` and in many cases `@.` 
and `@byrow` will give equivalent results. But there are important 
deviations in behavior. Consider the setup 

```julia
df = DataFrame(a = [1, 2], b = [3, 4])
```

* Control flow. In all versions of Julia, expressions of the form 
  `if...else`, `a ? b : c` cannot be broadcasted. In versions below
  1.7-dev, expressions of the form `a && b` and `a || b` cannot be 
  broadcasted. Consequently, the `@.` macro will fail when encountering such

>   control flow while `@byrow` will not. 

Author: re-written.

  ```
  julia> @with df @byrow begin 
             if :a == 1
                 5
             else 
                 10
             end
         end
  2-element Vector{Int64}:
    5
   10
  
  julia> @with df @. begin 
             if :a == 1
                 5
             else 
                 10
             end
         end # will error
  ```

* Broadcasting objects that are not columns. `@byrow` constructs an 
  anonymous function *which accepts only the columns of the dataframe*
  and broadcasts that function. Consequently, it does not broadcast
  objects that are referenced which are not columns. 
  
  ```julia
  julia> df = DataFrame(a = [1, 2], b = [3, 4]);
  julia> @with df @byrow :x + [5, 6]
  ```
  
  will error, because the `:x` in the above expression refers
  to a scalar `Int`, and you cannot do `1 + [5, 6]`.

  On the other hand

  ```julia
  @with df @. :x + [5, 6]
  ```

>   will not, as `df.x` is a 2-element vector as is `[5, 6]`, and 
>   `1 .+ [5, 6]` is allowed.

Reviewer: `1 .+ [5, 6]` is never called, right? The two arguments have the same length.
```suggestion
  will succeed, as `df.x` is a 2-element vector as is `[5, 6]`.
```

  Because `ByRow` inside `transform` blocks does not internally 
  use broadcasting in all circumstances, in the rare instance
  that a column in a data frame is a custom vector type that
  implements custom broadcasting, this custom behavior will 

>   be called with `@byrow`.

Reviewer: Is this what you mean?
```suggestion
  not be called with `@byrow`.
```

> * Broadcasting expensive calls. In Base Julia, broadcastsing 

Reviewer: ```suggestion
* Broadcasting expensive calls. In Base Julia, broadcasting 
```

  evaluates calls first and then broadcasts the result. Because
  `@byrow` constructs an anonymous function and evaluates
  that function for every row in the data frame, expensive functions
  will be evaluated many times. 
  
  ```julia
  julia> function expensive()
             sleep(.5)
             return 1
         end;
  
  julia> @time @with df @byrow :a + expensive();
    1.037073 seconds (51.67 k allocations: 3.035 MiB, 3.19% compilation time)
  
  julia> @time @with df :a .+ expensive();
    0.539900 seconds (110.67 k allocations: 6.525 MiB, 7.05% compilation time)
  
  ```
  

>   This problem comes up when using the `@.` macro as well, but can easily be   fixed with `$`.

Reviewer: ```suggestion
This problem comes up when using the `@.` macro as well, but can easily be fixed with `$`.
```
